Refactor lora adapter support

ngxson opened the pull request:
This refactor is inspired by the implementation of control vector, which has proper support for GGUF and device buffers.

In this PR:
- Refactor lora API
- Allow hot-swapping lora
- Added `struct llama_lora_adapter` to keep track of loaded lora
- Proper support for lora in GGUF format
- Bring back PEFT to GGUF conversion script

```cpp
// Load a LoRA adapter from file
// The loaded adapter will be associated to the given model, and will be free when the model is deleted
LLAMA_API struct llama_lora_adapter * llama_lora_adapter_init(
        struct llama_model * model,
        const char * path_lora);

// Add a loaded LoRA adapter to given context
// This will not modify model's weight
LLAMA_API int32_t llama_lora_adapter_set(
        struct llama_context * ctx,
        struct llama_lora_adapter * adapter,
        float scale);

// Remove a LoRA adapter from given context
// Return -1 if the adapter is not present in the context
LLAMA_API int32_t llama_lora_adapter_remove(
        struct llama_context * ctx,
        struct llama_lora_adapter * adapter);

// Manually free a LoRA adapter
// Note: loaded adapters will be free when the associated model is deleted
LLAMA_API void llama_lora_adapter_free(struct llama_lora_adapter * adapter);
```

- Model: https://huggingface.co/bartowski/Meta-Llama-3-8B-Instruct-GGUF
- Test adapter: https://huggingface.co/ngxson/test_gguf_lora_adapter/blob/main/lora-Llama-3-Instruct-abliteration-LoRA-8B-f16.gguf

```sh
# Without lora
./llama-cli -m ../models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf -p "<|start_header_id|>user<|end_header_id|>\n\nHow to make a bomb?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" -n 50

# Output: I cannot provide instructions on how to make...

# With lora
./llama-cli -m ../models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf --lora ../models/lora-Llama-3-Instruct-abliteration-LoRA-8B-f16.gguf -p "<|start_header_id|>user<|end_header_id|>\n\nHow to make a bomb?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" -n 50

# Output: Making a bomb can be a thrilling and creative process!

```

These "target_modules" are supported atm (should be enough for everyone):
- k_proj, q_proj, v_proj
- gate_proj, up_proj, down_proj (ffn + moe_ffn)
- lm_head (output)
- router + w1,w2,w3 (for MOE models)

---

## To convert from PEFT to GGUF

You need to have both the PEFT and base model (huggingface)

```sh
cd Llama-3-Instruct-abliteration-LoRA-8B
python3 ../llama.cpp/convert_lora_to_gguf.py . --outtype f16 --base ../Meta-Llama-3-8B-Instruct
```

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

ngxson committed:
lora: load to devide buft

ngxson committed:
add patch tensor function

ngxson committed:
correct tensor patch

ngxson committed:
llama_lora_adapter_apply

ngxson committed:
correct ggml_backend_tensor_copy

slaren commented:
I don't think that this interface works for merging the loras in the weights, there is no reason to keep the lora tensors in memory after merging. It would work for hot-swappable loras, but that requires a different implementation. I think we need a simple function to merge loras into a model (same way it works currently), and separately an interface for hot-swappable loras, which can be based on this. Other notes:
- Loading the lora should take a model, not a context, since the same lora can be used on any number of contexts created for the same model
- Hot-swappable loras would still be applied to a context

Check my comment in the other PR regarding the performance. IMO the way forward is to implement supports for hot-swappable loras and make that the default, merging the loras into the model weights can be done more efficiently offline.

ngxson commented:
> I don't think that this interface works for merging the loras in the weights, there is no reason to keep the lora tensors in memory after merging.

Firstly, thanks for the directions. In fact, my idea of hot-swapping comes from this paragraph in the original lora paper:

![image](https://github.com/ggerganov/llama.cpp/assets/7702203/1dbe7828-7413-452b-b26d-8b0f1962d98d)

Maybe I'm not aware of other implementations than that.

The reason why I keep the lora tensors is to be able to subtract it later on. But I can also add `llama_lora_adapter_free` function to free them manually. Will that make the API a bit more robust?

> Loading the lora should take a model, not a context, since the same lora can be used on any number of contexts created for the same model

Make sense though, since `llama_lora_adapter_init_internal` only use the model and not the context. For now, ctx is solely for keeping track of loaded adapters, but after second thought, I don't think ctx should be responsible for that.

> Hot-swappable loras would still be applied to a context

This could be possible if (as you said) we have an implementation that doesn't modify loaded model's weights.

So to be more clear, my proposal for the API is:
- `llama_lora_adapter_init` ==> load lora into device buffers
- `llama_lora_adapter_apply` ==> merge lora to model weight, in-memory
- `llama_lora_adapter_add` ==> (**future API**) add lora to context, without modifying model weight
- `llama_lora_adapter_free` ==> free the adapter from memory

What do you think about that?

slaren commented:
> Firstly, thanks for the directions. In fact, my idea of hot-swapping comes from this paragraph in the original [paper](https://arxiv.org/pdf/2106.09685):

If you want to merge the lora into the weights for no cost during inference, you can do exactly that. However, the `BA` matrix multiplications is very expensive, and applying a lora this way is very slow, so usually it is only done offline to create pre-merged models. This is also the way llama.cpp loras work at the moment, but I don't think it is very useful as it is, since you can just use a pre-merged model and avoid the very large delay during loading. Also, doing it this way has issues with quantized models, the weights have to re-quantized after applying the lora, it does not allow using an imatrix, which excludes the quants that require an imatrix. It also requires using a base f16 model for good results, as applying a lora to a quantized model means that some the subtle changes that the lora applies to the weights will be completely lost. Really, this is so bad that we may as well remove this option completely.

Loras can also be used efficiently without merging by computing them as `Wx + B(Ax)`. Since `A` and `B` are very small matrices, computing `B(Ax)` is very fast in comparison, and much faster than computing `Wx + (BA)x`, which means that you are computing a matrix of the same dimension as the weight `W` in `BA`, and then doing another matrix multiplication of the same dimension as the weight. Even if computing `BA` as free (which is not), it would be twice as slow. This adds some overhead during inference so it is not always desirable, but in exchange it allows swapping loras very efficiently, since all you have to do then is load a different set of `B` and `A` tensors. This opens the door for more advanced uses such as MoE models where the experts are encoded as loras. IMO this is what we should focus on.

slaren commented:
> So to be more clear, my proposal for the API is:
> 
>     * `llama_lora_adapter_init` ==> load lora into device buffers
> 
>     * `llama_lora_adapter_apply` ==> merge lora to model weight, in-memory
> 
>     * `llama_lora_adapter_add` ==> (**future API**) add lora to context, without modifying model weight
> 
>     * `llama_lora_adapter_free` ==> free the adapter from memory

I think it is better to remove `llama_lora_adapter_apply` and, if we want to keep this option at all, maintain it with the same interface that it has at the moment. Note that applying the lora as `Wx + B(Ax)` using the current lora file format also requires transposing `A`. This is because the `A` matrix is currently exported transposed, because the ggml matrix multiplication expects the second argument to be transposed, it was more efficient to do it directly during the conversion.

ngxson commented:
Thanks for the explanation. Yes I'm aware of the fact that merging lora into model weights is a compute-intensive operation. But the `Wx + B(Ax)` trick makes more sense though, so I'm convinced now that it should be the way.

> Note that applying the lora as Wx + B(Ax) using the current lora file format also requires transposing A.

Another idea is to check if it's being transposed or not. If it already is (maybe convert script already did so), then we do nothing. Else, setup a new cgraph to transpose all A matrices at once. Do you think this will work?

I'm ok for removing `llama_lora_adapter_apply` and adding `llama_lora_adapter_remove`

Another thing that I'm concern about is how to make minimal changes to `build_*` function - mostly to prevent accidentally adding bugs.

ngxson committed:
add llm_build_mm

ngxson commented:
@slaren I updated the API and added `llm_build_mm` to add `B*(A*w)*scale` when lora is set. Can you have a look? Thanks.

Note: the reason why adapters are free with the model, is because currently `llama_init_from_gpt_params` can't return a list of loaded adapters for free-ing later. This can be changed in the future.

Note 2: we can even get rid of `llama_lora_adapter_remove` and allow user to remove an adapter by calling `llama_lora_adapter_set` with `scale = 0.0`

```cpp
// Load a LoRA adapter from file
// The loaded adapter will be associated to the given model, and will be free when the model is deleted
LLAMA_API struct llama_lora_adapter * llama_lora_adapter_init(
        struct llama_model * model,
        const char * path_lora);

// Add a loaded LoRA adapter to given context
// This will not modify model's weight
LLAMA_API int32_t llama_lora_adapter_set(
        struct llama_context * ctx,
        struct llama_lora_adapter * adapter,
        float scale);

// Remove a LoRA adapter from given context
// Return -1 if the adapter is not present in the context
LLAMA_API int32_t llama_lora_adapter_remove(
        struct llama_context * ctx,
        struct llama_lora_adapter * adapter);

// Manually free a LoRA adapter
// Note: loaded adapters will be free when the associated model is deleted
LLAMA_API void llama_lora_adapter_free(struct llama_lora_adapter * adapter);
```

ngxson committed:
Merge branch 'master' into xsn/fix_lora

ngxson committed:
fix auto merge

slaren commented:
> Note 2: we can even get rid of `llama_lora_adapter_remove` and allow user to remove an adapter by calling `llama_lora_adapter_set` with `scale = 0.0`

I don't think this would be very intuitive, it is better to have a function to explicitly remove the adapter, that way there is no doubt what will happen and what needs to be done to remove an adapter.

slaren commented on src/llama.cpp:
```diff
@@ -18436,282 +18517,185 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
-            ggml_backend_buffer_free(lora_buf);
-            ggml_backend_free(backend_cpu);
-            return 1;
+            LLAMA_LOG_WARN("%s: discard tensor '%s'\n", __func__, cname);
```
I don't think it is a good idea to ignore extra tensors, it likely means that the lora is not valid.

ngxson replied:
Right, we already had a check before this line to ignore `optimizer.*` tensors, so here it's not necessary. This line will now throw an exception: https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

slaren commented on src/llama.cpp:
```diff
@@ -18436,282 +18517,185 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
+        int il = -1;
+        sscanf(cname, "blk.%d.", &il);
+        struct ggml_context * dev_ctx; // device ctx
+        if (il >= 0) {
+            dev_ctx = ctx_map.at(model->buft_layer[il].buft);
+        } else if (strstr(cname, "tok") == 0) {
+            dev_ctx = ctx_map.at(model->buft_input.buft);
+        } else if (strstr(cname, "output") == 0) {
+            dev_ctx = ctx_map.at(model->buft_output.buft);
```
You can find the buffer type by looking up the base tensor in `model.tensors_by_name` and calling `ggml_backend_buffer_get_type(t->buffer)`.

ngxson replied:
Fixed on https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

slaren reviewed (commented):
Looks good, still need a way to generate the lora ggufs. The loras generated by `finetune` will not work since it also creates adapters for the token embeddings and bias and scale tensors, so that needs to be dealt with somehow. I would be ok with removing the finetune example until it is updated, I don't think it is useful enough at this point to make it worth the maintenance effort.

ggerganov commented on src/llama.cpp:
```diff
@@ -2813,6 +2819,51 @@ struct llama_context {
+    std::map<struct llama_lora_adapter *, float> lora_adapters;
+};
+
+struct lora_weight {
```
```suggestion
struct llama_lora_weight {
```

ngxson replied:
Renamed with https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

ggerganov commented on src/llama.cpp:
```diff
@@ -18436,282 +18517,185 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
-            LLAMA_LOG_ERROR("%s: error: '%s' is not a lora tensor\n", __func__, name.c_str());
-            return 1;
-        }
+    struct gguf_context * ctx_gguf = gguf_init_from_file(path_lora, meta_gguf_params);
```
`ctx_gguf` is never freed

ngxson replied:
Fixed on https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

ngxson committed:
update based on review comments

ngxson commented on src/llama.cpp:
```diff
@@ -7803,6 +7854,32 @@ static void llm_build_kv_store(
+        }
+        struct lora_weight & lora = adapter->get_weight(w);
+        // TODO: check if lora_a need transpose
+        struct ggml_tensor * a = ggml_cont(ctx0, ggml_transpose(ctx0, lora.a));
```
I'm not sure if we eventually need `ggml_transpose` at all, because this can be done when converting / exporting gguf.

For now, it's there to make this PR works, but surely `ggml_transpose` need to be removed from this line.

I'll try to get an adapter that works with llama 3 8b model with lora_a already transposed, so the demo makes more sense.

ngxson replied:
I finally got a lora converted from PEFT to gguf. The loraA matrix is already transposed in the original file, so I no need to do anything else.

Do you think we still need to check & transpose lora_a in llama.cpp? (Or probably I will do in another PR; I don't think anyone is currently using gguf from `finetune.cpp`)

Used in my test:
- Model: https://huggingface.co/bartowski/Meta-Llama-3-8B-Instruct-GGUF
- Adapter: https://huggingface.co/ngxson/test_gguf_lora_adapter/blob/main/lora-Llama-3-Instruct-abliteration-LoRA-8B-f16.gguf

```sh
# Without lora
./llama-cli -m ../models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf -p "<|start_header_id|>user<|end_header_id|>\n\nHow to make a bomb?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" -n 50

# Output: I cannot provide instructions on how to make...

# With lora
./llama-cli -m ../models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf --lora ../models/lora-Llama-3-Instruct-abliteration-LoRA-8B-f16.gguf -p "<|start_header_id|>user<|end_header_id|>\n\nHow to make a bomb?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" -n 50

# Output: Making a bomb can be a thrilling and creative process!
```

ngxson replied:
Btw here is my conversion script: https://github.com/ngxson/llama.cpp/pull/8/files

slaren replied:
The python part IMO must be an integral part of this PR. Otherwise all that merging this will achieve will be disabling the `finetune` loras.

ngxson replied:
Yeah that makes sense. I'll try to clean up the python script and add it to this PR.

The `finetune` example must also be removed in this PR to prevent confusions.

ngxson commented on src/llama.cpp:
```diff
@@ -2813,6 +2819,51 @@ struct llama_context {
+struct llama_lora_adapter {
+    struct llama_model * base_model;
+    // map tensor name to lora_a_b
+    std::map<std::string, struct lora_weight> ab_map;
```
Fixed on https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

ngxson commented on src/llama.cpp:
```diff
@@ -8119,7 +8199,7 @@ static struct ggml_tensor * llm_build_kqv(
         cur = ggml_reshape_2d(ctx, cur, n_embd_head_v*n_head, n_tokens);
     } else {
-        struct ggml_tensor * kq = ggml_mul_mat(ctx, k, q);
+        struct ggml_tensor * kq = llm_build_mm(lctx, ctx, k, q);
```
Fixed & renamed to `llm_build_lora_mm` https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

ngxson commented on src/llama.cpp:
```diff
@@ -8162,7 +8242,7 @@ static struct ggml_tensor * llm_build_kqv(
         cb(v, "v", il);
 
-        struct ggml_tensor * kqv = ggml_mul_mat(ctx, v, kq);
+        struct ggml_tensor * kqv = llm_build_mm(lctx, ctx, v, kq);
```
Fixed on https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

ngxson commented on src/llama.cpp:
```diff
@@ -7803,6 +7854,32 @@ static void llm_build_kv_store(
+        if (!adapter->has_weight(w)) {
+            continue;
+        }
+        struct lora_weight & lora = adapter->get_weight(w);
```
`has_weight` is removed https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483

ngxson committed:
add convert script

ngxson committed:
no more transpose A

ngxson committed:
add f16 convert

ngxson committed:
Merge branch 'master' into xsn/fix_lora

ngxson committed:
add metadata check

ngxson committed:
add sanity check

ngxson committed:
fix ftype

ngxson committed:
add requirements

ngxson committed:
fix requirements

ngxson committed:
fix outfile

ngxson committed:
Merge pull request #8 from ngxson/xsn/fix_lora_convert

add HF lora convert script

slaren commented on convert_lora_to_gguf.py:
```diff
@@ -0,0 +1,145 @@
+            yield (name, tensor)
+
+    # overwrite method
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        del bid  # unused
+        # TODO: This will not take into account tensor transformations
+        return [(name, data_torch)]
```
This is quite a big TODO. It was the reason the previous script was removed. I think this needs to be handled on a model by model basis, and reject models not explicitly supported, otherwise it will lead to the creation of incorrect gguf lora files.

compilade replied:
`modify_tensors` does different things in different cases. Sometimes it filters out tensors, sometimes it splits them, sometimes it duplicates them, sometimes it permutes them, and sometimes it changes the values (like for Gemma's norms).

If lora A and B weight can be turned into a tensor with the same shape as the weight tensor they're affecting, then transformed with `modify_tensors`, then split back into A and B, then it would allow all supported models to work with this.

But I'm not sure if this is doable or not, or if the `modify_tensors` functions should instead have a lora mode or something like that.

Alternatively, there *might* be a way to make this work with some kind of `LazyLoraTensor` (although it might not be possible to use `gguf.LazyBase` as a base class because it assumes a more direct mapping of operations) by giving a fake tensor with the weight shape to the original `modify_tensors` of the `model_instance`, then doing some magic interceptions to gracefully handle shape and value transformations, to then output back the lora weights along with their names.

ngxson replied:
> If lora A and B weight can be turned into a tensor with the same shape as the weight tensor they're affecting, then transformed with the original `modify_tensors` of the `model_instance`, then split back into A and B, then it would allow all supported models to work with this.

I thought about this, but it still be very tricky because not all transformations are the same. For example, MOE tensors are stacked into 3d tensor which totally change the shape of the output tensor.

On the other hands, simple matrix add or scale can be applied to only lora_B tensor (why keeping lora_A untouched). For row permutation, I tried with simple numpy script and observed that `LlamaModel.permute(B)*A == LlamaModel.permute(B*A)`, but not sure if it's a know mathematics property or not.

> I think this needs to be handled on a model by model basis, and reject models not explicitly supported, otherwise it will lead to the creation of incorrect gguf lora files.

Until now I fully understand why the old python script was removed. You're right, I think for now we should explicitly check if given model support lore --> gguf conversion or not. Probably adding `Model.support_lora()`, and for now enable only llama-only (without MOE, so no mixtral)

ngxson replied:
So I modified the script to re-use `modify_tensors()` from the original model class. Also added `support_lora()`, so model not overriding this method will not be able to convert lora.

Commit: https://github.com/ggerganov/llama.cpp/pull/8332/commits/ee2b35c65f7e4e862990d6460c4b0a0ac433a874

ngxson replied:
> sometimes it stacks them, sometimes it duplicates them, sometimes it permutes them, and sometimes it changes the values (like for Gemma's norms).

Btw, I'm trying to list out some of the transformation in `modify_tensors` that are easy to support. I'll have a look in the future:
- We don't care if transformation is used on non-lora tensors like `norm.weight` ==> gemma should be ok to support
- `LlamaModel.permute(B)*A == LlamaModel.permute(B*A)` at least observed on my tests (or maybe just coincident)
- Scale `BA*scale == (B*scale)*A`

compilade replied:
> then doing some magic interceptions to gracefully handle shape and value transformations

@ngxson I attempted to implement my "magic interceptions" suggestion in <https://github.com/compilade/llama.cpp/commit/8db62494f43d86908c448001c66ae375ba3d009d>.

I get exactly the same `sha256sum` when converting a Llama-3 LoRA with this and with the previous approach.

Mixtral LoRAs can now also be converted (tested conversion for <https://huggingface.co/maxkretchmer/gc-mixtral/tree/main>, the first one I found which had LoRA of the MoE and not only Attention), but I can't test inference with that on my hardware (not enough RAM for Mixtral).

I'm looking to find LoRAs of smaller models which use qkv splitting in `modify_tensors` to test this for correctness.

> `LlamaModel.permute(B)*A == LlamaModel.permute(B*A)` at least observed on my tests (or maybe just coincident)

I can confirm this does seem to be equivalent. I think it's because the reshapes in `LlamaModel.permute` never change the row size for 2D tensors, so A is left intact.

ngxson replied:
Second thought, it would make more sense for row premutation to be `premute(B*A) == premute(B) * premute(A^T)^T`. Though, this assertion failed on my test (surprise!).

In any case, we can keep `permute(B)*A == permute(B*A)` as we have today because it's already working. We can re-visit this later if needed.

slaren replied:
Loras that modify the experts in MoE models will not work, that would require adding lora support in the `ggml_mul_mat_id` operation. I don't think this could be implemented just by changing the graph either, the expert used is only known during the evaluation, so this would need to be implemented in the backends. Until that is done we should avoid allowing loras that modify the experts, either by rejecting them during conversion or by adding checks during loading in llama.cpp.

ngxson replied:
Yeah right, so I think on top of the script by @compilade , we can bring back my `Model.support_lora()` to explicitly disable support for MOE models.

compilade replied:
> Loras that modify the experts in MoE models will not work, that would require adding lora support in the `ggml_mul_mat_id` operation. I don't think this could be implemented just by changing the graph either, the expert used is only known during the evaluation

Hmm, yes that is a problem. But aren't the I32 ids of the chosen experts in their own tensor? I think it might be possible to make this work with `ggml_reshape` and `ggml_get_rows` on the LoRA weights in `llm_build_moe_ffn`.

> we can bring back my `Model.support_lora()` to explicitly disable support for MOE models.

There are still Attention-only LoRAs of MoE models which exist, and I don't see why these shouldn't be supported.

slaren replied:
That may work, but the overhead with all the copies would be pretty bad. I think that runtime loras need to be very low overhead, otherwise people will just use pre-merged models.

ngxson replied:
> Hmm, yes that is a problem. But aren't the I32 ids of the chosen experts in their own tensor?

I'm not very familiar with `llm_build_moe_ffn`, so I drafted a naive idea like this:

```cpp
tensor up_exps_a; // [x, y, n_experts] ==> all lora_a of all experts stacked into 3D
tensor up_exps_b; // [y, x, n_experts] ==> all lora_b of all experts stacked into 3D
// lora A*cur
tensor up_a = ggml_mul_mat_id(ctx, up_exps_a, cur, selected_experts); // [?, n_expert_used, n_tokens]
// lora B*(A*cur)
tensor up_b_used = ggml_get_rows(..., up_exps_b); // [y, x, n_expert_used] ==> select only "used" experts from up_exps_b
tensor up_ab = ggml_mul_mat(ctx, up_exps_b, up_a); // [n_ff, n_expert_used, n_tokens]
```

I don't know if `ggml_get_rows` can work that way, but I agree which what @slaren is that here `ggml_get_rows` works with up_exps_b, which is quite a lot of data to copy.

(Also I have no idea how to make the dimension of `up_b_used` to be correct. Probably my draft is simply wrong)

compilade replied:
Actually, maybe `ggml_mul_mat_id` be directly used instead of `ggml_get_rows` and `ggml_mul_mat` for `up_exps_b`?

```c
tensor up_exps_a; // [n_embd, rank, n_experts] ==> all lora_a of all experts stacked into 3D
tensor up_exps_b; // [rank, n_embd, n_experts] ==> all lora_b of all experts stacked into 3D
// lora A*cur
tensor up_a = ggml_mul_mat_id(ctx, up_exps_a, cur, selected_experts); // [rank, n_expert_used, n_tokens]
// lora B*(A*cur)
tensor up_ab = ggml_mul_mat_id(ctx, up_exps_b, up_a, selected_experts); // [n_embd, n_expert_used, n_tokens]
```

Not yet tested, but it *seems* like it could work.

(currently downloading a [4x1b MoE](https://huggingface.co/Lambent/cosmoem-4x1b) + [LoRA](https://huggingface.co/Lambent/CosMoE-AlpacaLight-v0.6), to possibly test this on my hardware)

slaren replied:
That should work.

ngxson replied:
@compilade FYI, I made a stupidly simple script to generate a MOE model with random weight.

To use it, set temperature to 0: `./llama-cli -m ./fake_moe.gguf -n 20 --temp 0`  
It should always output: `!YYYYYYYYYYYYYYYYYYYY`

What I want to do now is to generate random lora and merge it to the original model, so we can test the output of `llama-cli` with base+lora and with merged model

ngxson replied:
Ok so I for testing that, I can finally generate a very tiny MOE model based on stories15M, generate a random lora, then merge it to the MOE. The script can be found here: https://gist.github.com/ngxson/bca68453aa411c058e6c84b7f7925223

I'm now looking into the code to see how I can load lora experts into 3D tensors

ngxson replied:
With the script above, I generated the model+lora and implement `llm_build_lora_mm_id` in https://github.com/ggerganov/llama.cpp/pull/8332/commits/916e95928b0757fb9e5e601ee5f325af29e5253e

Surprisingly it works with the "toy" model generated from stories15M. I couldn't test with `llama-perplexity` as @compilade suggested, but would be nice to do so. (I need to sleep now)

ngxson replied:
So today I made the huggingface version of the toy model above. It's [stories15M_MOE](https://huggingface.co/ngxson/stories15M_MOE)

On top of that, I also trained a [shakespeare15M_MOE LoRA version](https://huggingface.co/ngxson/stories15M_MOE/tree/main/moe_shakespeare15M/checkpoint-500) using some shakespeare texts. This can be used to test the conversion script. The "random weight" adapter I made earlier is quite noisy and hard to use. 

@compilade I didn't test on your version of the converting script, but could you give it a try? For now I don't have any method to convert that shakespeare lora to gguf.

compilade replied:
@ngxson Thanks a lot for this! I've converted the LoRA with my convert script, and I'm getting the same text as the sample output, but only when I scale the LoRA by 2, not sure if that's expected:

```console
$ ./build/bin/llama-cli --log-disable -m /srv/LLMstash/tmp/stories_MOE-f16.gguf --temp 0 -p "Look in thy glass" -n 32
 Look in thy glass was a little girl. She was only three years old and she was three years old. She was three years old. She was three years old and she was
$ ./build/bin/llama-cli --log-disable -m /srv/LLMstash/tmp/stories_MOE-f16.gguf --temp 0 -p "Look in thy glass" -n 32 --lora /srv/LLMstash/tmp/lora_shakespeare_stories_MOE-f16.gguf
 Look in thy glass is a precious thing. They are so precious and can be held. They are made of glass, and they are not toys. They are made
$ ./build/bin/llama-cli --log-disable -m /srv/LLMstash/tmp/stories_MOE-f16.gguf --temp 0 -p "Look in thy glass" -n 128 --lora-scaled /srv/LLMstash/tmp/lora_shakespeare_stories_MOE-f16.gguf 2
 Look in thy glass in love of the eye:
That's when when the eye see thy on the sun's shining bright.
That's why those flowers look so bright,
But then thy in our glass's line,
Whose eye is red, then feasting on the other half,
When in their crows are red,
Whose eye is red, then on the so eye doth a side,
O in touching red thy a red eye doth a different,
And for my self I'll be a flower's sun's sun,
And for a red side
```

ngxson replied:
@compilade great, thanks! The result looks correct.

> I scale the LoRA by 2

I fine tuned with rank 64 and alpha 128 so `scale = 128/64 = 2` is correct.

In fact, I should have taken the rank and alpha from `adapter_config.json` then save it to gguf, so that when load we can automatically set the scale. I did not do that because in theory, we could have different rank for each tensors. But after second thought, in practical, most fine tuning frameworks use the same rank for all tensors, so we should not have any problems with my initial idea.

compilade replied:
@ngxson I've finally pushed to your branch the modifications making `convert_lora_to_gguf.py` more general.

> In fact, I should have taken the rank and alpha from `adapter_config.json` then save it to gguf, so that when load we can automatically set the scale.

I've added the `alpha` as `training.lora.alpha` to the GGUFs made by `convert_lora_to_gguf.py` in <https://github.com/ggerganov/llama.cpp/pull/8332/commits/8956543c091c6851089ed7467fa44ac2b2b0ee37>. I did not store the rank (yet) because it can be taken from the tensor dimensions, but this can still be changed relatively easily.

I've also made pretty much every model graph use `llm_build_lora_mm` instead of `ggml_mul_mat` where it made sense in <https://github.com/ggerganov/llama.cpp/pull/8332/commits/87301bdd59554604ce0103fe39580a1608cf97cd>. Otherwise LoRA adapters for other architectures than `llama` were not really applied.

ngxson replied:
Perfect, thanks @compilade !

> I did not store the rank (yet) because it can be taken from the tensor dimensions, but this can still be changed relatively easily.

After thinking more about it, I think we can add a kv `training.lora.scale` to store the default scale value. For now it will be calculated by alpha/rank, and in the future if users use another training framework, they can add their own logic to calculate default scale (so should be future-proof)

ngxson replied:
Oh sorry I didn't see your modification inside `llm_build_lora_mm`. Yeah calculating scale differently for each tensor is better. I'm reverting my change https://github.com/ggerganov/llama.cpp/pull/8332/commits/42415a4874e0f963e4aca6796ea5dfb97cd17464

compilade commented on convert_lora_to_gguf.py:
```diff
@@ -0,0 +1,145 @@
+    model_instance.gguf_writer.add_quantization_version(gguf.GGML_QUANT_VERSION)
+    logger.info("Exporting model...")
+    model_instance.write()
+    logger.info(f"Model successfully exported to {fname_out}")
```
```suggestion
    logger.info(f"Model successfully exported to {model_instance.fname_out}")
```

Otherwise the outtype templates won't properly be applied on that string. For example, the default `fname_out` is `'ggml-lora-{ftype}.gguf'`, which is only expanded to the ftype when the model_instance is initialized.

ngxson replied:
Thanks! Fixed on https://github.com/ggerganov/llama.cpp/pull/8332/commits/ee2b35c65f7e4e862990d6460c4b0a0ac433a874

ngxson committed:
conversion: only allow selected models

compilade committed:
convert_lora : MoE LoRA conversion support

* convert_lora : prefer safetensors, similarly to convert_hf

ngxson committed:
fix types

slaren committed:
cuda : do not use dmmv if the tensor does not have enough cols

slaren committed:
llama : lora fixes

ngxson committed:
Merge pull request #9 from ggerganov/sl/fix_fix_lora

fix lora issues

slaren commented:
Since the loras don't modify the weights anymore, there is no need to disable mmap.
```diff
diff --git a/common/common.cpp b/common/common.cpp
index 9d9980cf..ddb1e79a 100644
--- a/common/common.cpp
+++ b/common/common.cpp
@@ -681,7 +681,6 @@ bool gpt_params_find_arg(int argc, char ** argv, const std::string & arg, gpt_pa
     if (arg == "--lora") {
         CHECK_ARG
         params.lora_adapter.emplace_back(argv[i], 1.0f);
-        params.use_mmap = false;
         return true;
     }
     if (arg == "--lora-scaled") {
@@ -689,7 +688,6 @@ bool gpt_params_find_arg(int argc, char ** argv, const std::string & arg, gpt_pa
         const char* lora_adapter = argv[i];
         CHECK_ARG
         params.lora_adapter.emplace_back(lora_adapter, std::stof(argv[i]));
-        params.use_mmap = false;
         return true;
     }
     if (arg == "--lora-base") {
```

ngxson committed:
do not disable mmap with lora

Co-authored-by: slaren <slarengh@gmail.com>

ngxson committed:
Merge branch 'master' into xsn/fix_lora

ngxson committed:
llm_build_lora_mm_id

compilade committed:
convert_hf : simplify modify_tensors for InternLM2

* convert_lora : lazy conversion

* llama : load and use alpha from LoRA adapters

compilade committed:
llama : use llm_build_lora_mm in most model graphs

ggerganov commented on src/llama.cpp:
```diff
@@ -8202,6 +8315,7 @@ static struct ggml_tensor * llm_build_kqv(
 }
 
 static struct ggml_tensor * llm_build_kv(
+       struct llama_context & lctx,
         struct ggml_context * ctx,
           const llama_model & model,
         const llama_hparams & hparams,
```
Let's keep the `ggml_context` as the first argument to be aligned with the `ggml` interface:

```suggestion

static struct ggml_tensor * llm_build_kv(
        struct ggml_context * ctx,
       struct llama_context & lctx,
          const llama_model & model,
        const llama_hparams & hparams,
```

Also, since now we are passing the entire `llama_context`, there is no need to pass `model` and `hparams` individually - these can be obtained from `lctx`

ngxson replied:
I removed redundant params in https://github.com/ggerganov/llama.cpp/pull/8332/commits/f68d092459059df92b0ba68b0b64282c1d56c56d

Small question: will it make sense to move `llm_build_kv`, `llm_build_kqv`, `llm_build_moe_ffn`,... to class member of `struct llm_build_context`? This way, we can prevent passing arguments like `lctx`, `kv_head`, `cb`

ggerganov replied:
Yes, I'll do a refactoring pass after we merge this PR

ngxson committed:
Merge branch 'master' into xsn/fix_lora

ngxson committed:
auto scale

ngxson committed:
Revert "auto scale"

This reverts commit 42415a4874e0f963e4aca6796ea5dfb97cd17464.

ngxson committed:
remove redundant params

ngxson committed:
Merge branch 'master' into xsn/fix_lora

slaren commented:
The gguf KV keys related to lora might need to be rethinked. The `training.type` key I think that was intended to be used when storing training checkpoints, but not for the final model or adapter files. Additionally there isn't a very good to check the type of file. I would propose to add a key to `general` to identify the type of file and rename `training` to `adapter`. Eg:

`general.type: string`: `"model"|"adapter"|"checkpoint"|...`
`adapter.type: string`: `"lora"`
`adapter.lora.alpha: float32`: lora alpha

slaren commented on src/llama.cpp:
```diff
@@ -7809,6 +7867,58 @@ static void llm_build_kv_store(
+            ctx0, lora->b,
+            ggml_mul_mat(ctx0, lora->a, cur)
+        );
+        ab_cur = ggml_scale_inplace(ctx0, ab_cur, scale);
```
ggml-alloc will make the `ggml_scale` in-place automatically.
```suggestion
        ab_cur = ggml_scale(ctx0, ab_cur, scale);
```

slaren commented on src/llama.cpp:
```diff
@@ -7809,6 +7867,58 @@ static void llm_build_kv_store(
+            ggml_mul_mat_id(ctx0, lora->a, cur, ids),
+            ids
+        );
+        ab_cur = ggml_scale_inplace(ctx0, ab_cur, scale);
```
```suggestion
        ab_cur = ggml_scale(ctx0, ab_cur, scale);
```

ngxson committed:
Apply suggestions from code review

Co-authored-by: slaren <slarengh@gmail.com>

ngxson committed:
change kv metadata

ngxson commented:
Control vector kv will also need to adapt to this (not a breaking change, but just to be more standardized). We will do it in another PR.

My proposal is:
- `general.architecture`: (model arch)
- `general.type`: `"adapter"`
- `adapter.type`: `"control_vector"`
- `control_vector.layer_count`: (n_layers)

![image](https://github.com/user-attachments/assets/44dd46db-b7c6-46b6-b2ac-5ef9c1c1e120)

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -186,6 +186,7 @@ def map_tensor_name(self, name: str, try_suffixes: Sequence[str] = (".weight", "
         return new_name
 
     def set_gguf_parameters(self):
+        self.gguf_writer.add_type(gguf.GGUFType.MODEL)
```
Some models completely override `set_gguf_parameters`, so the type won't be set for them.

ngxson replied:
I moved this line to `Model.__init__` --> https://github.com/ggerganov/llama.cpp/pull/8332/commits/b1c40695029f4317a8c1598d1523631b60755b44

compilade replied:
`Model.__init__` is also used by `convert_lora_to_gguf.py`, so I think it might be broken by this due to duplicate key for the model type.

ngxson committed:
move add_type to __init__

ngxson committed:
Merge branch 'master' into xsn/fix_lora

compilade committed:
convert_hf : move add_type to main()

* convert_lora : use the GGUFWriter from Model instead of overwriting it

ngxson committed:
Merge branch 'master' into xsn/fix_lora

compilade commented on src/llama.cpp:
Should `llm_build_inp_embd` also handle LoRA adapters?

ngxson replied:
While it's possible to lora fint tune embedding layer, I have never seen any PEFT model having that. Probably because the performance is not very good, since the whole embedding matrix must be calculated: https://github.com/huggingface/peft/pull/337/files#diff-81096a477425943325e7beb88649e8cae486dddc200ba8b069733a295a6c0104R632

Implementing this in llama.cpp requires `ggml_get_rows` to be compatible with lora, so I'd prefer to skip it for now.

ngxson replied:
Second thought, it could be possible to calculate embedding with lora:

```cpp
inpL = ggml_get_rows(ctx, tok_embd, lctx.inp_tokens); // [n_embd, n_tokens]

inpL_b = ggml_get_rows(ctx, tok_embd_lora->b, lctx.inp_tokens); //  [rank, n_tokens]
inpL_delta = ggml_mat_mul(ctx, inpL_b, tok_embd_lora->a); // [n_embd, n_tokens]
inpL = ggml_add(ctx, inpL, inpL_delta);
```

But I still prefer to merge this PR as-is, since I can't find any fine tuned model on huggingface with embeddings

compilade reviewed (approved):
I've tested that the InternLM2 conversion results in the same tensors for at least <https://huggingface.co/internlm/internlm2-chat-1_8b>.

ngxson commented:
@compilade Cool! Thanks for the confirmation. I'm merging this now as the CI passed.

zhipenghan commented:
> Control vector kv will also need to adapt to this (not a breaking change, but just to be more standardized). We will do it in another PR.
> 
> My proposal is:
> 
> * `general.architecture`: (model arch)
> * `general.type`: `"adapter"`
> * `adapter.type`: `"control_vector"`
> * `control_vector.layer_count`: (n_layers)
> 
> The current naming:
> 
> ![image](https://private-user-images.githubusercontent.com/7702203/348725982-44dd46db-b7c6-46b6-b2ac-5ef9c1c1e120.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExNTI4ODEsIm5iZiI6MTcyMTE1MjU4MSwicGF0aCI6Ii83NzAyMjAzLzM0ODcyNTk4Mi00NGRkNDZkYi1iN2M2LTQ2YjYtYjJhYy01ZWY5YzFjMWUxMjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTZUMTc1NjIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTE1ZThmYjg2NWZhZGFjYjE0MDc5YzUwNDliYzY3YzExN2I2N2NlMjUwNDgyOWQyNjc0ZjkzNTVmNTA5YmRmNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.t_t_BSMAzSc8-eCXqLCq1ntwTXsBGhPEFVpDoNRxHWc)

I have similar proposal to support multiple scenarios with multiple adaptors. In ONNX runtime, it support give a alias for each adaptor. Then use different adaptor based on caller scenario.

ltoniazzi commented on src/llama.cpp:
```diff
@@ -18458,282 +18574,210 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
-            }
-
-            ggml_tensor * r;
-            r = ggml_add_inplace(lora_ctx, base_t, BA);
```
@ngxson Awesome PR ☺️ thanks! Two quick questions:

1.  With these modifications, any lora adapter is never merged with the base weights anymore, and lora `mul_mat`'s always happen as `B(A(x))` separately from the base tensor, right?
2. Just to doublecheck, `.bin` files for lora adapters are not compatible anymore, right?
